test/validation/api/ipsec/ipsec_test_out.c

```diff
@@ -1371,6 +1371,53 @@ static void ipsec_test_capability(void)
 	CU_ASSERT(odp_ipsec_capability(&capa) == 0);
 }
 
+static void ipsec_test_default_values(void)
+{
+	odp_ipsec_config_t config;
+	odp_ipsec_sa_param_t sa_param;
+
+	memset(&config, 0x55, sizeof(config));
+	memset(&sa_param, 0x55, sizeof(sa_param));
+
+	odp_ipsec_config_init(&config);
+	CU_ASSERT(config.inbound.lookup.min_spi == 0);
+	CU_ASSERT(config.inbound.lookup.max_spi == UINT32_MAX);
+	CU_ASSERT(config.inbound.lookup.spi_overlap == 0);
+	CU_ASSERT(config.inbound.retain_outer == ODP_PROTO_LAYER_NONE);
+	CU_ASSERT(config.inbound.parse_level == ODP_PROTO_LAYER_NONE);
+	CU_ASSERT(config.inbound.chksums.all_chksum == 0);
+	CU_ASSERT(config.outbound.all_chksum == 0);
+	CU_ASSERT(!config.stats_en);
+
+	odp_ipsec_sa_param_init(&sa_param);
+	CU_ASSERT(sa_param.proto == ODP_IPSEC_ESP);
+	CU_ASSERT(sa_param.crypto.cipher_alg == ODP_CIPHER_ALG_NULL);
```

Comment:
Just wondering what should be the behavior if the implementation doesn't support NULL algos.

Reply:
TL; DR: We can leave things as they are or consider changing the default authentication algorithm to SHA-1.

odp_ipsec_sa_param_init() is supposed to set the default values as specified in the API spec regardless of the actual algorithm support in the implementation. As you must have noticed, this PR does not change the default algorithms but just adds a test for what was already required.

If an implementation does not support null encryption or null authentication, we have a situation where the default parameters set by odp_ipsec_sa_param_init() cannot be passed to odp_ipsec_sa_create() and the application has to set the parameters to suitable values. But that is not much different from parameters which do not have default value defined.

That said, null encryption in ESP is likely supported by many implementations because it is mandatory in IPsec and also tends to be somewhat easier to implement than most other ciphers.

Null authentication is no longer mandatory in IPsec and even if it is supported, it may not be combined with null encryption in real usage (but in performance tuning such combo can be useful and ODP API allows it).

Maybe SHA-1, which is mandatory in IPsec and thus likely supported by most implementations, would be a better default value for the authentication algorithm than the current null authentication.

Reply:
> If an implementation does not support null encryption or null authentication, we have a situation where the default 
> parameters set by odp_ipsec_sa_param_init() cannot be passed to odp_ipsec_sa_create() and the application has to set
> the parameters to suitable values.

This is what I had in mind. If application is expected to check crypto capabilities before doing odp_ipsec_sa_create(), then we can keep existing behavior. RFC compliance would be the rationale to state algos as mandatory. But implementations need not be RFC compliant. 

Other option is to leave default values for algos as implementation defined. Would that be fine? But then testing such a field would be very complicated.

Reply:
> If application is expected to check crypto capabilities before doing odp_ipsec_sa_create()

Yes, it is expected to do that or accept that the creation may fail.

> Other option is to leave default values for algos as implementation defined. Would that be fine?

I think that would be ok too, but I do not think it would be any more useful to an application than the null defaults. You cannot just let the implementation choose the crypto algorithms for an SA and you would have to give keys of correct length too.

Note that this PR is not changing the default values for the algorithms but only adds a validation check for what was already specified in the API.

Reply:
> You cannot just let the implementation choose the crypto algorithms for an SA and you would have to give keys of
> correct length too

Yeah. In that sense, it is better to retain the current behavior (with NULLs).

> Note that this PR is not changing the default values for the algorithms but only adds a validation check for what was
> already specified in the API.

Agreed. I was just trying to understand the spec requirements in case NULL algos are not supported.